src/snowflake/snowpark/mock/_functions.py

```diff
@@ -1700,42 +1700,53 @@ def mock_initcap(values: ColumnEmulator, delimiters: ColumnEmulator):
 
 @patch("convert_timezone")
 def mock_convert_timezone(
-    target_timezone: ColumnEmulator,
-    source_time: ColumnEmulator,
-    source_timezone: Optional[ColumnEmulator] = None,
+    *args: ColumnEmulator,
 ) -> ColumnEmulator:
     """Converts the given source_time to the target timezone.
 
     For timezone information, refer to the `Snowflake SQL convert_timezone notes <https://docs.snowflake.com/en/sql-reference/functions/convert_timezone.html#usage-notes>`_
     """
     import dateutil
 
-    is_ntz = source_time.sf_type.datatype.tz is TimestampTimeZone.NTZ
-    if source_timezone is not None and not is_ntz:
-        SnowparkLocalTestingException.raise_from_error(
-            ValueError(
+    # mock_convert_timezone matches the sql function call semantics.
+    # It has different parameters when called with 2 or 3 args.
+    # When called with two args, the third will be replaced with None.
+    if args[2] is None:
+        target_timezone, source_time, _ = args
+        source_timezone = pandas.Series([None] * len(source_time))
+        return_type = TimestampTimeZone.TZ
+    else:
+        source_timezone, target_timezone, source_time = args
+        return_type = TimestampTimeZone.NTZ
+        if source_time.sf_type.datatype.tz is not TimestampTimeZone.NTZ:
+            raise ValueError(
                 "[Local Testing] convert_timezone can only convert NTZ timestamps when source_timezone is specified."
             )
-        )
 
-    # Using dateutil because it uses iana timezones while pytz would use Olson tzdb.
-    from_tz = None if source_timezone is None else dateutil.tz.gettz(source_timezone)
+    combined = pandas.concat(
+        [source_timezone, target_timezone, source_time], axis=1, ignore_index=True
+    )
 
-    if from_tz is not None:
-        timestamps = [ts.replace(tzinfo=from_tz) for ts in source_time]
-        return_type = TimestampTimeZone.NTZ
-    else:
-        timestamps = list(source_time)
-        return_type = TimestampTimeZone.TZ
+    def _convert(row):
+        source_timezone, target_timezone, source_time = row
+        if source_time is None:
+            return None
 
-    res = []
-    for tz, ts in zip(target_timezone, timestamps):
-        # Add local tz if info is missing
-        if ts.tzinfo is None:
-            ts = LocalTimezone.replace_tz(ts)
+        if source_timezone is not None:
+            # Using dateutil because it uses iana timezones while pytz would use Olson tzdb.
+            source_time = source_time.replace(tzinfo=dateutil.tz.gettz(source_timezone))
+
+        if source_time.tzinfo is None:
+            source_time = LocalTimezone.replace_tz(source_time)
+
+        result = source_time.astimezone(dateutil.tz.gettz(target_timezone))
+
+        if return_type == TimestampTimeZone.NTZ:
+            result = result.replace(tzinfo=None)
```

Comment:
this logic seems new, is this also fixing a bug that previously we don't remove tzinfo TimestampNTZ type?

Reply:
That is correct.